src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb

```diff
@@ -303,6 +301,147 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+            Else
+                TupleTypeSymbol.VerifyTupleTypePresent(elements.Length, node, Me.Compilation, diagnostics)
+            End If
+
+            Return New BoundTupleLiteral(node, elementNamesArray, boundArguments.ToImmutableAndFree(), tupleTypeOpt, hasErrors)
+        End Function
+
+        ''' <summary>
+        ''' Returns the type to be used as a field type generates errors in case the type is not
+        ''' supported for tuple type fields.
+        ''' </summary>
+        Private Function GetTupleFieldType(expression As BoundExpression, errorSyntax As VisualBasicSyntaxNode, diagnostics As DiagnosticBag, <Out> ByRef hasError As Boolean) As TypeSymbol
+            Dim expressionType As TypeSymbol = expression.Type
+
+            If Not expression.HasErrors Then
+                If expression.Type IsNot Nothing Then
+                    If expressionType.SpecialType = SpecialType.System_Void Then
+                        expressionType = ErrorTypeSymbol.UnknownResultType
+                        hasError = True
+                        ReportDiagnostic(diagnostics, errorSyntax, ERRID.ERR_VoidValue)
+
+                    ElseIf (expressionType.IsRestrictedType()) Then
+                        hasError = True
+                        ReportDiagnostic(diagnostics, errorSyntax, ERRID.ERR_RestrictedType1, expressionType)
+
+                    End If
+                End If
+            End If
+
+            Return expressionType
+        End Function
+
+        Private Shared Sub CollectTupleFieldMemberNames(name As String, position As Integer, tupleSize As Integer, ByRef elementNames As ArrayBuilder(Of String))
+            ' add the name to the list
+            ' names would typically all be there Or none at all
+            ' but in case we need to handle this in error cases
+            If elementNames IsNot Nothing Then
+                elementNames.Add(If(name, TupleTypeSymbol.TupleMemberName(position)))
+            Else
+                If name IsNot Nothing Then
+                    elementNames = ArrayBuilder(Of String).GetInstance(tupleSize)
+                    For j As Integer = 1 To position - 1
+                        elementNames.Add(TupleTypeSymbol.TupleMemberName(j))
+                    Next
+                    elementNames.Add(name)
+                End If
+            End If
+        End Sub
+
+        Private Shared Function CheckTupleMemberName(name As String, position As Integer, syntax As VisualBasicSyntaxNode, diagnostics As DiagnosticBag, uniqueFieldNames As PooledHashSet(Of String)) As Boolean
+            Dim reserved As Integer = TupleTypeSymbol.IsElementNameReserved(name)
+            If reserved = 0 Then
+                Binder.ReportDiagnostic(diagnostics, syntax, ERRID.ERR_TupleReservedMemberNameAnyPosition, name)
+                Return False
+
+            ElseIf reserved > 0 AndAlso reserved <> position + 1 Then
```

Comment:
In that case, consider renaming `position` to `index` to follow the convention @jcouv suggested.